llvm/lib/Transforms/IPO/ArgumentPromotion.cpp

```diff
@@ -610,6 +639,45 @@ static bool findArgParts(Argument *Arg, const DataLayout &DL, AAResults &AAR,
       // unknown users
     }
 
+    auto *CB = dyn_cast<CallBase>(V);
+    Value *PtrArg = cast<Value>(U);
+    if (CB && PtrArg && CB->getCalledFunction() == CB->getFunction()) {
+      Type *PtrTy = PtrArg->getType();
+      APInt Offset(DL.getIndexTypeSizeInBits(PtrTy), 0);
+      PtrArg = PtrArg->stripAndAccumulateConstantOffsets(
+          DL, Offset,
+          /* AllowNonInbounds= */ true);
+      if (PtrArg != Arg)
+        return false;
+
+      if (Offset.getSignificantBits() >= 64)
+        return false;
+
+      int64_t Off = Offset.getSExtValue();
+      if (Off) {
```

Comment:
Calling stripAndAccumulateConstantOffsets() just to check for a zero offset afterwards is pretty pointless. Use stripPointerCasts() instead -- or just don't strip at all, now that we have opaque pointers.

Reply:
I get your point that strips are anyways not there in opaque ptrs, but how to collect offsets ? I couldn't find a function which will accumulate constant offsets on a Value! Please take a look at testcase recursion-non-zero.ll submitted with this PR. We are trying to check if a ptr is being offset at some non-zero offset.

Reply:
How do you propose I check zero offset ? because checking zero offset is a necessary check and it it's not we need to bail out in this optimization.

stripAndAccumulateConstantOffsets() doesn't do much, it is just traversing the GEP tree and appending the constant offsets to Offset variable.

Reply:
You check for zero offset using `PtrArg == Arg`.

Reply:
@nikic the function https://github.com/llvm/llvm-project/blob/e4163c0927772f2ec73cf16d53e000614c419c45/llvm/lib/IR/Value.cpp#L731 pretty much runs a do-while loop and traverses the GEP tree to accumulate the offsets. If it encounters a bitcast (which it won't), it doesn't do much. So, I would want to ideally pull out just the traversal code out from the function, skipping the bitcasts code. But I think that can another patch, is it blocking this one ?